### ISSUE TITLE
Highcharts.cleanRecursively should not recursively process class objects.

### ISSUE THREAD

**Issue by marklofdahl** [highcharts/highcharts#14346]
#### Expected behaviour
Highcharts.cleanRecursively should call isClass(), in order to skip class objects. This is the behavior of Highcharts.merge.

#### Actual behaviour
Highcharts.cleanRecursively does not call isClass(), while Highcharts.merge does call isClass();

#### Product version
Hicharts 8.2

#### Affected browser(s)
All

**Comment by pawelfus** [comment]
Thanks for reporting the issue @marklofdahl 

Could you provide live demo which shows the issue in action? What does it break etc. We will check it.

Thanks!

**Comment by marklofdahl** [comment]
It was causing infinite recursion in Highcharts.cleanRecursively. I will have to find some time to make a reproducible example, probably next week sometime. Not always quick to distill that out of a large code base, as I'm sure you have encountered ;-)

**Comment by pawelfus** [comment]
Yes, I know :) I would start with checking what's passed to `chart.update()`. Perhaps the config is incorrect?

**Comment by marklofdahl** [comment]
Ok, figured it out more and reproduced it.

bad behavior, using a class object: http://jsfiddle.net/marklofdahl/tsyL95ma - causes infinite recursion.
good behavior, using a simple object: http://jsfiddle.net/marklofdahl/2rdtcyse

My object has a reference to something returned by chart.renderer.text. This item has recursive references to itself.

The problem is because Highcharts.merge DOES NOT recursively copy class-objects, but Highcharts.cleanRecursively DOES recursively evaulate class-objects.

If I use a simple-object instead of a class-object, it works.
If I change Highcharts.cleanRecursively to not recursively evaluate class-objects, it works.
If I don't store references to renderer objects (or anything that has recursive references to itself), it works, although it is doing a lot of extra compares for things I don't need it to.

The background for this is that I have an object I need access to inside callback functions (like label formatters, etc). I pass my object as part of the config, and I can access it with chart.userOptions.myObj.
Is there a proper way to pass my object in the config in such a way that Highcharts.merge and Highcarts.cleanRecursively will just leave it alone and not process it recursively?

Thanks.

**Comment by pawelfus** [comment]
Thank you for the details @marklofdahl 

I'm still not convinced if it's a bug - the config with a class is not a valid options object. In theory `fooObj` is not valid either. However, we never discouraged to use options that way, so perhaps a simple change might resolve this.

Anyway, for now I would use something like this:
```js
const createChart = (fooObj, options) => {
  const chart = Highcharts.chart('container', options);

  chart.fooObj = fooObj;

  return chart;
}
```

So later, instead of `chart.update({ fooObj })`, you simply change it like this: `chart.fooObj = fooObj;`.

I'm leaving the decision up to @TorsteinHonsi

**Comment by marklofdahl** [comment]
Thanks for the response. I tried using the createChart approach you suggested. However, some of my callback functions (which need fooObj to be set) are executed in the rendering process before Highcharts.chart has completed, so I can't set chart.fooObj after the call has completed.
However, I now understand it enough that I have a couple other ways to change my code or otherwise work around it, so I should be fine.
Whatever you decide with Highcharts.cleanRecursively will be ok.
Thanks.
